NetKVM/Common/ParaNdis_TX.cpp

```diff
@@ -1100,3 +1100,53 @@ ULONG CNB::Copy(PVOID Dst, ULONG Length) const
 
     return Copied;
 }
+
+ULONG CNB::Copy(PVOID Dst, ULONG Length, ULONG DataOffset) const
+{
+    /* This function searches for a required offset within a chain of MDLs iteratively.
+     * Chosen for rapid deployment, this interim solution is not fully optimized and
+     * may lead to performance penalties with long MDL chains. Future revisions should
+     * aim for optimization to mitigate any performance concerns.
+     */
+    ULONG Offset = 0;
+    PMDL DataOffsetMDL = nullptr;
+
+    for (DataOffsetMDL = NET_BUFFER_CURRENT_MDL(m_NB);; DataOffsetMDL = DataOffsetMDL->Next)
+    {
+        Offset += MmGetMdlByteCount(DataOffsetMDL);
+        if (Offset > DataOffset + NET_BUFFER_CURRENT_MDL_OFFSET(m_NB) || DataOffsetMDL->Next == nullptr)
+        {
+            break;
+        }
+    }
```

Comment:
How the procedure is protected from getting too big DataOffset parameter (> than summary length of MDL chain)?
In such case the loop finishes as   DataOffsetMDL->Next is NULL and the Offset is still small.
I think in such case it should return 0. What will it do?

Reply:
indeed, I have now added boundary protection and resubmitted the code

Reply:
I suggest to do that different way and do not rewrite this procedure anymore. This is the last thing in this series that I do not feel satisfied from. As an idea:

static CopyFromMdlChain(PVOID Dest, ULONG Len, PMDL& Source, ULONG& Offset /* in MDL */)
{
  while (Source && MmGetMdlByteCount(Source) < Offset) {
     Offset -= MmGetMdlByteCount(Source);
     Source = Source->Next;
  }
  if (!Source) return 0;
  then copy in a loop, always advance Source and recalculate the offset
  return copied;
}

And the caller do something like that:

ULONG len = GetDataLength() - ParsedHeadersLength;
PMDL mdl = NET_BUFFER_CURRENT_MDL(m_NB);
ULONG offset = ParsedHeadersLength + NET_BUFFER_CURRENT_MDL_OFFSET(m_NB);

for (all the pages until len > 0) {
  ULONG toCopyNow = lastPage ? len : PAGE_SIZE
  done = CopyFromMdlChain(dest, len, mdl, offset);
  if (not good) break and process error
  dest+= toCopyNow;
  len -= toCopyNow;
}
Such a way we start from the first MDL only once. What you think?

Reply:
@ybendito Okay, I have made the changes and resubmitted.